contributors/design-proposals/service-injection-policy.md

```diff
@@ -0,0 +1,444 @@
+service-by-service basis, without the pod authors having to incorporate the
+information into every pod spec where it is needed.
+
+## Constraints and Assumptions
+
+1.  Future work might require new mechanisms to be made to work with controllers
+    such as deployments and replicasets that create pods. Existing controllers
+    that create pods should recreate their pods when a new Service Injection
+    Policy is added that would effect them.
+
+## Use Cases
+
+1.  As a user, I want to be able to describe a way that pods should be injected
+    with the information to consume a particular service. This would allow me
+    to concisely model the information about how the service
+    should be consumed without altering every consuming pod spec.
+
+### Loose coupling between services and their consumers
+
+Using a Service Injection Policy allows pod template authors to not have to explicitly
+set information for every pod. This way authors of pod templates consuming a
+specific service do not need to know all the details about that service.
+
+The information will be injected based off the policy's `Selector`.
+
+## Proposed Changes
+
+### ServiceInjectionPolicy API object
+
+This resource is alpha. The policy itself is immutable.
+
+```go
+type ServiceInjectionPolicy struct {
+    unversioned.TypeMeta
+    ObjectMeta
+
+    Spec ServiceInjectionPolicySpec
+}
+
+type ServiceInjectionPolicySpec struct {
+    // Selector defines which Pods should be modified, this only targets Pods.
+    Selector      unversioned.LabelSelector
+    // Env defines the collection of EnvVar to inject.
+    Env           []EnvVar `json:omitempty`
+    // EnvFrom defines the collection of EnvFromSource to inject.
+    EnvFrom       []EnvFromSource `json:omitempty`
+    // Volumes defines the collection of Volume to inject.
+    Volumes       []Volume `json:omitempty`
+    // VolumeMounts defines the collection of VolumeMount to inject.
+    VolumeMounts  []VolumeMount `json:omitempty`
+}
+```
+
+#### Validations
+
+In order for the Service Injection Policy to be valid it must fulfill the
+following constraints:
+
+- The `Selector` field must be defined. This is how we know which pod templates
+  to inject so therefore it is required.
+- The policy must define _at least_ 1 of `Env`, `EnvFrom`, `Volumes`, or
+  `VolumeMounts`.
+- The `Volume`s given in `Volumes` must exist and be valid.
+
+#### Conflicts
+
+There are a number of edge conditions that might occur at the time of
+injection. These are as follows:
+
+- Merging lists with no conflicts: if a pod already has a volume defined
+  exactly as defined in the SIP. This would replace the other volume.
+- Merging lists with conflicts: if a SIP redefines an `EnvVar` or a `Volume`,
+  the original will be replaced and an annotation to the Pod spec will reflect
+  a warning documenting the decision.
+- Conflicts between `Env` and `EnvFrom`: this would throw an error.
```

Comment:
When we write this up in the docs it might be good to make a more generic statement too - for example: in the event of conflict between the PodSpec and a PIP, the PIP wins.